### PR TITLE
Add registerUserSettingsChange handler

### PR DESCRIPTION
**Pull request by vivek3003**
The following PR adds a new registerHandler.

This handler can be used to register for changes in user's settings. The callback will be triggered whenever the settings subscribed to change.

Currently, adding support for file open preference.

Sample usage:

```
microsoftTeams.registerUserSettingsChangeHandler([microsoftTeams.userSettingKeys.fileOpenPreference], (settingKey, newValue) => {
        // The setting (identified by the key has been changed by the user
    });
```

### CODE REVIEW

File: src/internal/globalVars.ts [hunk 1]
```diff
@@ -46,4 +46,5 @@ export class GlobalVars {
   public static remoteCameraErrorHandler: (error: remoteCamera.ErrorReason) => void;
   public static remoteCameraDeviceStateChangeHandler: (deviceStateChange: remoteCamera.DeviceState) => void;
   public static remoteCameraSessionStatusChangeHandler: (sessionStatusChange: remoteCamera.SessionStatus) => void;
+  public static userSettingsChangeHandler: (updatedSettingKey: userSettingKeys, updatedValue: any) => void;
```
Comment (shishir-srivastava) [inline src/internal/globalVars.ts:49]:
```suggestion
  public static userSettingsChangeHandler: (settingKey: userSettingKeys, value: any) => void;
```
The handler is already called a "change handler". I feel that adding the "updated" prefix to each param may be redundant.

File: src/internal/internalAPIs.ts [hunk 2]
```diff
@@ -116,6 +117,12 @@ function handleAppButtonHoverLeave(): void {
   }
 }
 
+function handleUserSettingsChange(udatedSettingKey, updatedSettingValue): void {
+  if (GlobalVars.userSettingsChangeHandler) {
+    GlobalVars.userSettingsChangeHandler(udatedSettingKey, updatedSettingValue);
+  }
+}
```
Comment (shishir-srivastava) [inline src/internal/internalAPIs.ts:124]: Same as above (param naming).

File: src/public/interfaces.ts [hunk 3]
```diff
@@ -167,6 +167,15 @@ export enum FileOpenPreference {
   Web = 'web',
 }
 
+/**
+ * User setting changes that can be subscribed to,
+ */
+export enum userSettingKeys {
```
Comment (shishir-srivastava) [inline src/public/interfaces.ts:173]: This should be in Pascal case.

Comment (shishir-srivastava) [reply src/public/interfaces.ts:173]: Also, add provision to subscribe to theme changes.

**Review by sachingupta — CHANGES_REQUESTED** [review]
We just merge one of major refactoring PR which will impact the way we register handler. https://github.com/OfficeDev/microsoft-teams-library-js/pull/518

Please merge latest master and update your changes.

**Comment by vivek3003** [comment]
> We just merge one of major refactoring PR which will impact the way we register handler. #518
> 
> Please merge latest master and update your changes.

@sachingupta Updated the PR.

File: src/internal/handlers.ts [hunk 4]
```diff
@@ -88,6 +90,14 @@ export function registerBeforeUnloadHandler(handler: (readyToUnload: () => void)
   handler && sendMessageToParent('registerHandler', ['beforeUnload']);
 }
 
+export function registerUserSettingsChangeHandler(
+  settingKeys: UserSettingKeys[],
+  handler: (settingKey: UserSettingKeys, value: any) => void,
```
Comment (sachingupta) [inline src/internal/handlers.ts:95]: We should use registerHandler, which ca be extended to take params/args as part of register handler.

File: src/public/publicAPIs.ts [hunk 5]
```diff
@@ -328,3 +329,17 @@ export function initializeWithFrameContext(
...
   setFrameContext(frameContext);
 }
+
+/**
+ * register a handler to be called when a user setting changes. The changed setting key & value is provided in the callback.
+ * @param settingKeys List of user setting changes to subscribe
+ * @param handler When a subscribed setting is updated this handler is called
+ */
+export function registerUserSettingsChangeHandler(
+  settingKeys: UserSettingKeys[],
```
Comment (sachingupta) [inline src/public/publicAPIs.ts:339]: We just need to extend registerHandler to support args in this case settingKeys and use it here.

**Review by sachingupta — CHANGES_REQUESTED** [review]

File: src/private/interfaces.ts [hunk 6]
```diff
@@ -43,6 +43,23 @@ export enum ViewerActionTypes {
   editNew = 'editNew',
 }
 
+/**
+ * * @private
+ * Hide from docs.
+ * ------
+ * User setting changes that can be subscribed to,
+ */
+export enum UserSettingKeys {
```
Comment (sachingupta) [inline src/private/interfaces.ts:52]: May be we just say UserSettingTypes ? Keys looks very technical here.

File: src/private/privateAPIs.ts [hunk 7]
```diff
@@ -203,3 +204,18 @@ export function getConfigSetting(callback: (value: string) => void, key: string)
...
+
+/**
+ * @private
+ * register a handler to be called when a user setting changes. The changed setting key & value is provided in the callback.
+ * @param settingKeys List of user setting changes to subscribe
+ * @param handler When a subscribed setting is updated this handler is called
+ */
+export function registerUserSettingsChangeHandler(
+  settingKeys: UserSettingKeys[],
+  handler: (settingKey: UserSettingKeys, value: any) => void,
```
Comment (sachingupta) [inline src/private/privateAPIs.ts:216]: with this design we are expecting handler to execute for each setting change. Which make sense.

**Review by sachingupta — APPROVED** [review]

**Review by charpercyr — APPROVED** [review]

File: src/private/privateAPIs.ts [hunk 8]
```diff
@@ -203,3 +204,18 @@ export function getConfigSetting(callback: (value: string) => void, key: string)
...
+): void {
+  ensureInitialized(FrameContexts.content);
```
Comment (ydogandjiev) [inline src/private/privateAPIs.ts:218]: This API should work in any frame context so you probably shouldn't pass a frame context to the `ensureInitialized` call.

**Review by ydogandjiev — APPROVED** [review]

File: src/private/privateAPIs.ts (hunk 8 above)
Comment (vivek3003) [reply src/private/privateAPIs.ts]: This is done

File: src/private/interfaces.ts (hunk 6 above)
Comment (vivek3003) [reply src/private/interfaces.ts]: Changed to UserSettingTypes.

**Pull request merged**